### ISSUE TITLE
Do not crash the app if uncaught exception is thrown?

### ISSUE THREAD

**Issue by nyarian** [tari-project/wallet-android#560]
By default, if an uncaught exception is caught, the application terminates and OS implementation-dependent strategy for application failure takes place (for pure Android a dialog is shown - https://developer.android.com/topic/performance/images/crash-example-framed.png).

It is possible to handle uncaught exceptions gracefully, by showing, for instance, an application dialog, describing that some error has happened, with an option to send error logs through the email and with the second option ("OK") relaunching the home page or the authentication page (as if the application was just launched). By this the user won't feel completely disconnected from the app, as the interaction session will be preserved (instead of complete application termination that forces user to relaunch the app IF he will want to after the crash).

**Comment by Tas4tari** [comment]
@kukabi To discuss with mobile team

**Comment by alexandrVakhtinTari** [comment]
It's not possible in terms of Android and iOS. It would be crashed anyway, we can just handle the crash and send it into sentry or crashlitycs.

**Comment by nyarian** [comment]
On iOS it's tricky indeed, as Swift's exception handling is very much unlike the strategy Objective-C uses, and NSUncaughtExceptionHandler gets invoked after the application is already crashed.

On Android it is very much possible, though, even if the crash happens on main thread in terms of Looper. I currently have it on a project I'm working on - works neat :)

**Comment by alexandrVakhtinTari** [comment]
I see, it definitely is. https://stackoverflow.com/questions/27829955/android-handle-application-crash-and-start-a-particular-activity

**Comment by nyarian** [comment]
Beware, though: overriding the default uncaught exception handler is only enough to recover from crashes happened in the background thread. Crash in the main thread will then result in ANR, and here's why:

1. Exception raised in the main thread.
2. Stack is unwinded up to the `Looper#loop`.
3. Looper dies.
4. Stack is unwinded to the root `main` procedure.
5. Uncaught exception handler is invoked.

The issue lies in the fact that the Looper is dead by the time the exception is caught by the handler, so the app renders invalid to continue any execution and will be stuck until Android detects it is like that and handles this fact appropriately (ANR dialog).

So, basically, if you are sure that the app might crash only in background thread, and all the main thread's invocations are related only to properly tested UI logic, then the solution with uncaught exception handler override will suffice. If the main thread is also a subject of uncaught exceptions handling, then there is a funny solution that covers both: 
![image](https://user-images.githubusercontent.com/34913433/151029800-87747623-fb70-4a43-8aab-4c2100793cf2.png)
